homeassistant/components/elmax/common.py

```diff
@@ -0,0 +1,229 @@
+"""Elmax integration common classes and utilities."""
+from __future__ import annotations
+
+from collections.abc import Mapping
+from datetime import timedelta
+import logging
+from logging import Logger
+from typing import Any
+
+import async_timeout
+from elmax_api.exceptions import (
+    ElmaxApiError,
+    ElmaxBadLoginError,
+    ElmaxBadPinError,
+    ElmaxNetworkError,
+)
+from elmax_api.http import Elmax
+from elmax_api.model.endpoint import DeviceEndpoint
+from elmax_api.model.panel import PanelEntry, PanelStatus
+
+from homeassistant.components.elmax.const import DEFAULT_TIMEOUT, DOMAIN
+from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ElmaxCoordinator(DataUpdateCoordinator):
+    """Coordinator helper to handle Elmax API polling."""
+
+    def __init__(
+        self,
+        hass: HomeAssistantType,
+        logger: Logger,
+        username: str,
+        password: str,
+        panel_id: str,
+        panel_pin: str,
+        name: str,
+        update_interval: timedelta,
+    ) -> None:
+        """Instantiate the object."""
+        self._client = Elmax(username=username, password=password)
+        self._panel_id = panel_id
+        self._panel_pin = panel_pin
+        self._panel_entry = None
+        self._state_by_endpoint = None
+        super().__init__(
+            hass=hass, logger=logger, name=name, update_interval=update_interval
+        )
+
+    @property
+    def panel_entry(self) -> PanelEntry | None:
+        """Return the panel entry."""
+        return self._panel_entry
+
+    @property
+    def panel_status(self) -> PanelStatus | None:
+        """Return the last fetched panel status."""
+        return self.data
+
+    def get_endpoint_state(self, endpoint_id: str) -> DeviceEndpoint | None:
+        """Return the last fetched status for the given endpoint-id."""
+        if self._state_by_endpoint is not None:
+            return self._state_by_endpoint.get(endpoint_id)
+        return None
+
+    @property
+    def http_client(self):
+        """Return the current http client being used by this instance."""
+        return self._client
+
+    async def _async_update_data(self):
+        try:
+            async with async_timeout.timeout(DEFAULT_TIMEOUT):
+                # Retrieve the panel online status first
+                panels = await self._client.list_control_panels()
+                panels = list(filter(lambda x: x.hash == self._panel_id, panels))
```

Comment:
We prefer a list comprehension instead of filter + lambda.

Reply:
Although since we're only interested in the first panel we can even use `next` and a generator expression. Then we avoid creating a list.
```py
panel = next((panel for panel in panels if panel.hash == self._panel_id), None)
if not panel:
    raise ConfigEntryAuthFailed(
        f"Panel ID {self._panel_id} is no more linked to this user account"
    )
```

Reply:
Done in #61273